sklearn/utils/estimator_checks.py

```diff
@@ -3050,6 +3055,126 @@ def check_class_weight_balanced_linear_classifier(name, Classifier):
     )
 
 
+@ignore_warnings(category=FutureWarning)
+def check_interaction_of_class_and_sample_weight_excluding_class(name, estimator_orig):
+    """Setting a class weights to zero is equivalent to excluding the samples
+    associated to this class from the calibration even when using non uniform
+    sample weights.
+    """
+    # Note: this test is similar to
+    # check_interaction_of_class_and_sample_weight_excluding_samples.
+    rng = np.random.RandomState(42)
+    err_msg_sw = (
+        f"For {name}, using class_weight as zero to one class is not "
+        "equivalent to making the sample_weight for the samples from that "
+        "class equals zero."
+    )
+    err_msg_exclude = (
+        f"For {name}, using class_weight as zero to one class is not "
+        "equivalent to excluding the samples from that class."
+    )
+
+    for n_classes in range(3, 10):
+        X, y = make_classification(
+            n_samples=200,
+            n_classes=n_classes,
+            n_informative=2 * n_classes,
+            random_state=0,
+        )
+        y = _enforce_estimator_tags_y(estimator_orig, y)
+        sample_weight = rng.uniform(size=y.shape[0])
+
+        # Model using `class_weight` with 0 weight for class 0
+        class_weight_dict = {cls: 1 if cls != 0 else 0 for cls in range(n_classes)}
+        estimator_cw = clone(estimator_orig).set_params(class_weight=class_weight_dict)
+        set_random_state(estimator_cw, random_state=0)
+        estimator_cw.fit(X, y, sample_weight=sample_weight)
+
+        # Model excluding samples from class 0
+        X_exclude = X[y != 0]
+        y_exclude = y[y != 0]
+        sample_weight_exclude = sample_weight[y != 0]
+        estimator_exclude = clone(estimator_orig)
+        set_random_state(estimator_exclude, random_state=0)
+        estimator_exclude.fit(X_exclude, y_exclude, sample_weight=sample_weight_exclude)
+
+        # Model setting `sample_weight` to 0 for samples from class 0
+        sample_weight_zero_weight_first = np.where(y == 0, 0, sample_weight)
+        estimator_sw = clone(estimator_orig)
+        set_random_state(estimator_sw, random_state=0)
+        estimator_sw.fit(X, y, sample_weight=sample_weight_zero_weight_first)
+
+        # Checking if the output is the same for multiple outputs
+        for method in ["predict", "predict_proba", "decision_function", "transform"]:
+            if hasattr(estimator_orig, method):
+                pred_cw = getattr(estimator_cw, method)(X)
+
+                pred_sw = getattr(estimator_sw, method)(X)
+                assert_allclose_dense_sparse(pred_cw, pred_sw, err_msg=err_msg_sw)
+
+                pred_exclude = getattr(estimator_exclude, method)(X)
+                if method in ["predict_proba", "decision_function"]:
+                    pred_cw = pred_cw[:, list(range(1, n_classes))]
+                    if method == "decision_function" and n_classes == 3:
+                        pred_cw = pred_cw[:, 1]
```

Comment:
Can you explain this, please?

Reply:
Because we excluded one class, for `n_classes` equal to 3, we are in a binary classification setting. The decision function outputs only one class output (similar to how you can get `predict_proba` of one class as the complement of another). Because of this, we have to exclude one column of the 3-class output to compare the arrays.

If I leave it as is I get the error:
```python
E           AssertionError:
...
E           (shapes (200, 2), (200,) mismatch)
E            x: array([[ 0.103091, -0.103091],
E                  [ 0.582987, -0.582987],
E                  [ 1.004391, -1.004391],...
E            y: array([-0.103091, -0.582987, -1.004391,  0.340705,  0.262283,  0.570671,
E                  -0.042541, -0.337614, -0.307904, -0.036075, -0.441613,  0.261148,
E                  -0.196478, -0.808903, -0.903717,  0.097252, -1.368786, -0.059082,...
```

I'm adding a comment to explain this.